File: day14/solution.js
let fs = require("fs");
let _ = require("lodash");

let data = fs.readFileSync("./data").toString();

function parseDataLine(line) {
	let name = line.split(" ")[0];
	let kmPerSec = parseInt(/(\d+) km\/s/.exec(line)[1], 10);
	let speedTime = parseInt(/km\/s for (\d+) seconds/.exec(line)[1], 10);
	let restTime = parseInt(/rest for (\d+) seconds/.exec(line)[1], 10);

	return { name, kmPerSec, speedTime, restTime };
}

function getDistanceAtTime(time, reindeerObject) {
	let kmPerSec = reindeerObject.kmPerSec;
	let speedTime = reindeerObject.speedTime;
	let restTime = reindeerObject.restTime;

	let distanceTraveled = 0;

	let timeSpeeding = 0;
	let timeResting = 0;
	let isResting = false;

	for (let i = 1; i <= time; i++) {
		if (timeSpeeding === speedTime) {
			isResting = true;
			timeSpeeding = 0;
		} else if (timeResting === restTime) {
			isResting = false;
			timeResting = 0;
		}

		if (!isResting) {
			distanceTraveled += kmPerSec;
			timeSpeeding++;
		} else {
			timeResting++;
		}
	}

	return distanceTraveled;
}

let seconds = 2503;
let reindeers = data.split("\n")
	.map(line => parseDataLine(line));

function getDistanceObject(reindeerObject, seconds) {
	return { name: reindeerObject.name, distance: getDistanceAtTime(seconds, reindeerObject) };
}

function inLeadAtSecond(reindeers, second) {
	let distanceObjects = reindeers
		.map(reindeerObject => getDistanceObject(reindeerObject, second));

	distanceObjects = distanceObjects.sort((a, b) => {
		return a.distance - b.distance;
	}).reverse();

	if (distanceObjects[0].distance === distanceObjects[1].distance) {
		return [distanceObjects[0], distanceObjects[1]];
	}

	return [distanceObjects[0]];
}

let reindeerNames = {};

reindeers.forEach(reindeer => {
	reindeerNames[reindeer.name] = 0;
});

for (let i = 1; i <= seconds + 1; i++) {
	let inLeads = inLeadAtSecond(reindeers, i);

	for (let inLead of inLeads)
		reindeerNames[inLead.name]++;
}

console.log("Part 1:", inLeadAtSecond(reindeers, seconds)[0].distance);
let kvPairs = Object.keys(reindeerNames).map(key => [key, reindeerNames[key]]);
console.log("Part 2:", kvPairs.sort((a, b) => a[1] - b[1]).reverse()[0][1]);

File: day14/solution.py
import operator
import re

data = open("data", "r").read()

def parseDataLine(line):
	name, kmPerSec, speedTime, restTime = re.search(
		r'(\w+) .* (\d+) km/s for (\d+) .* (\d+)', line).groups()

	kmPerSec = int(kmPerSec)
	speedTime = int(speedTime)
	restTime = int(restTime)

	return ( name, kmPerSec, speedTime, restTime )

def getDistanceAtTime(time, reindeerObject):
	name, kmPerSec, speedTime, restTime = reindeerObject

	distanceTraveled = 0

	timeSpeeding = 0
	timeResting = 0
	isResting = False
	for _ in range(time):
		if timeSpeeding == speedTime:
			isResting = True
			timeSpeeding = 0
		elif timeResting == restTime:
			isResting = False
			timeResting = 0

		if isResting:
			timeResting += 1
		else:
			distanceTraveled += kmPerSec
			timeSpeeding += 1

	return ( distanceTraveled, reindeerObject )


def inLeadAtSecond(reindeers, second):
	distanceObjects = map(lambda reindeerObj: getDistanceAtTime(second, reindeerObj), reindeers)

	sortedDistObj = list(reversed(sorted(distanceObjects)))

	if sortedDistObj[0] == sortedDistObj[1]:
		return [sortedDistObj[0][0], sortedDistObj[1][0]]

	return [sortedDistObj[0]]

seconds = 2503
reindeers = map(parseDataLine, data.split("\n"))

reindeerNames = {}
for reindeer in reindeers:
	reindeerNames[reindeer[0]] = 0

for second in range(seconds + 1):
	inLeads = inLeadAtSecond(reindeers, second)

	for inLead in inLeads:
		reindeerNames[inLead[1][0]] += 1

a = inLeadAtSecond(reindeers, seconds)[0]
print "Part 1, in lead:", (a[1][0], a[0])
print "Part 2, in lead:", list(reversed(sorted(reindeerNames.items(), key=operator.itemgetter(1))))[0]
